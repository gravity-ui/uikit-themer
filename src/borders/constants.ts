import type {BordersOptions} from './types.js';

export const DEFAULT_BORDERS: BordersOptions = {
    xs: '3px',
    s: '5px',
    m: '6px',
    l: '8px',
    xl: '10px',
};
