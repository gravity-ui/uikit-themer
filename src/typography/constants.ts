import type {TypographyOptions} from './types.js';

export const DEFAULT_TYPOGRAPHY_OPTIONS: TypographyOptions = {
    fontImports: [],
    fontFamilies: {
        sans: {
            mainFont: 'Inter',
            fallbackFonts: ['Helvetica Neue', 'Helvetica', 'Arial', 'sans-serif'],
        },
        monospace: {
            mainFont: 'Menlo',
            fallbackFonts: [
                'Monaco',
                'Consolas',
                'Liberation Mono',
                'Ubuntu Mono',
                'DejaVu Sans Mono',
                'Courier New',
                'Courier',
                'monospace',
            ],
        },
    },
    variants: {
        'body-short': {
            'font-size': '13px',
            'line-height': '16px',
        },
        'body-1': {
            'font-size': '13px',
            'line-height': '18px',
        },
        'body-2': {
            'font-size': '15px',
            'line-height': '20px',
        },
        'body-3': {
            'font-size': '17px',
            'line-height': '24px',
        },
        'caption-1': {
            'font-size': '9px',
            'line-height': '12px',
        },
        'caption-2': {
            'font-size': '11px',
            'line-height': '16px',
        },
        'code-1': {
            'font-size': '12px',
            'line-height': '18px',
        },
        'code-inline-1': {
            'font-size': '12px',
            'line-height': '14px',
        },
        'code-2': {
            'font-size': '14px',
            'line-height': '20px',
        },
        'code-inline-2': {
            'font-size': '14px',
            'line-height': '16px',
        },
        'code-3': {
            'font-size': '16px',
            'line-height': '24px',
        },
        'code-inline-3': {
            'font-size': '16px',
            'line-height': '20px',
        },
        'display-1': {
            'font-size': '28px',
            'line-height': '36px',
        },
        'display-2': {
            'font-size': '32px',
            'line-height': '40px',
        },
        'display-3': {
            'font-size': '40px',
            'line-height': '48px',
        },
        'display-4': {
            'font-size': '48px',
            'line-height': '52px',
        },
        'header-1': {
            'font-size': '20px',
            'line-height': '24px',
        },
        'header-2': {
            'font-size': '24px',
            'line-height': '28px',
        },
        'subheader-1': {
            'font-size': '13px',
            'line-height': '18px',
        },
        'subheader-2': {
            'font-size': '15px',
            'line-height': '20px',
        },
        'subheader-3': {
            'font-size': '17px',
            'line-height': '24px',
        },
    },
    groups: {
        body: {
            'font-family': 'sans',
            'font-weight': 400,
        },
        caption: {
            'font-family': 'sans',
            'font-weight': 400,
        },
        code: {
            'font-family': 'monospace',
            'font-weight': 400,
        },
        display: {
            'font-family': 'sans',
            'font-weight': 600,
        },
        header: {
            'font-family': 'sans',
            'font-weight': 600,
        },
        subheader: {
            'font-family': 'sans',
            'font-weight': 600,
        },
    },
};
