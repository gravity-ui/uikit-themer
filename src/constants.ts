import {DEFAULT_BORDERS} from './borders/constants.js';
import type {BaseColors, GravityTheme, Theme, UtilityColors} from './types.js';
import {DEFAULT_TYPOGRAPHY_OPTIONS} from './typography/constants.js';
import {generatePrivateColorsForBaseColors, replaceReferencesInUtilityColors} from './utils.js';

const DEFAULT_BRAND_COLOR = 'rgb(203,255,92)';

const DEFAULT_BACKGROUND = {
    dark: 'rgb(34,29,34)',
    light: 'rgb(255,255,255)',
};

const TEXT_CONTRAST_COLORS: Record<Theme, {white: string; black: string}> = {
    dark: {
        white: 'rgb(255, 255, 255)',
        black: 'rgba(0, 0, 0, 0.9)', // --g-color-private-black-900
    },
    light: {
        white: 'rgb(255, 255, 255)',
        black: 'rgba(0, 0, 0, 0.85)', // --g-color-private-black-850
    },
};

export const DEFAULT_BASE_COLORS: BaseColors = {
    white: {
        dark: {value: 'rgb(255, 255, 255)'},
        light: {value: 'rgb(255, 255, 255)'},
    },
    black: {
        dark: {value: 'rgb(0, 0, 0)'},
        light: {value: 'rgb(0, 0, 0)'},
    },
    brand: {
        dark: {value: DEFAULT_BRAND_COLOR},
        light: {value: DEFAULT_BRAND_COLOR},
    },
    orange: {
        dark: {value: 'rgb(200, 99, 12)'},
        light: {value: 'rgb(255, 119, 0)'},
    },
    green: {
        dark: {value: 'rgb(91, 181, 87)'},
        light: {value: 'rgb(59, 201, 53)'},
    },
    yellow: {
        dark: {value: 'rgb(255, 203, 0)'},
        light: {value: 'rgb(255, 219, 77)'},
    },
    red: {
        dark: {value: 'rgb(232, 73, 69)'},
        light: {value: 'rgb(255, 4, 0)'},
    },
    blue: {
        dark: {value: 'rgb(82, 130, 255)'},
        light: {value: 'rgb(82, 130, 255)'},
    },
    'cool-grey': {
        dark: {value: 'rgb(96, 128, 156)'},
        light: {value: 'rgb(107, 132, 153)'},
    },
    purple: {
        dark: {value: 'rgb(143, 82, 204)'},
        light: {value: 'rgb(143, 82, 204)'},
    },
};

const DEFAULT_PRIVATE_COLORS = generatePrivateColorsForBaseColors(
    DEFAULT_BASE_COLORS,
    DEFAULT_BACKGROUND.light,
    DEFAULT_BACKGROUND.dark,
);

const DEFAULT_UTILITY_COLORS: UtilityColors = {
    'base-brand': {
        dark: {value: DEFAULT_BRAND_COLOR},
        light: {value: DEFAULT_BRAND_COLOR},
    },
    'base-background': {
        dark: {value: DEFAULT_BACKGROUND.dark},
        light: {value: DEFAULT_BACKGROUND.light},
    },
    'base-brand-hover': {
        dark: {value: 'private.brand.650-solid'},
        light: {value: 'private.brand.600-solid'},
    },
    'base-selection': {
        dark: {value: 'private.brand.150'},
        light: {value: 'private.brand.200'},
    },
    'base-selection-hover': {
        dark: {value: 'private.brand.200'},
        light: {value: 'private.brand.300'},
    },
    'line-brand': {
        dark: {value: 'private.brand.600-solid'},
        light: {value: 'private.brand.600-solid'},
    },
    'text-brand': {
        dark: {value: 'private.brand.600-solid'},
        light: {value: 'private.brand.700-solid'},
    },
    'text-brand-heavy': {
        dark: {value: 'private.brand.700-solid'},
        light: {value: 'private.brand.700-solid'},
    },
    'text-brand-contrast': {
        dark: {value: TEXT_CONTRAST_COLORS.dark.black},
        light: {value: TEXT_CONTRAST_COLORS.light.black},
    },
    'text-link': {
        dark: {value: 'private.brand.550-solid'},
        light: {value: 'private.brand.600-solid'},
    },
    'text-link-hover': {
        dark: {value: 'private.brand.700-solid'},
        light: {value: 'private.brand.800-solid'},
    },
    'text-link-visited': {
        dark: {value: 'private.purple.700-solid'},
        light: {value: 'private.purple.550-solid'},
    },
    'text-link-visited-hover': {
        dark: {value: 'private.purple.850-solid'},
        light: {value: 'private.purple.800-solid'},
    },
};

export const DEFAULT_THEME: GravityTheme = {
    baseColors: DEFAULT_BASE_COLORS,
    privateColors: DEFAULT_PRIVATE_COLORS,
    utilityColors: replaceReferencesInUtilityColors(DEFAULT_UTILITY_COLORS, DEFAULT_PRIVATE_COLORS),
    typography: DEFAULT_TYPOGRAPHY_OPTIONS,
    borders: DEFAULT_BORDERS,
};
