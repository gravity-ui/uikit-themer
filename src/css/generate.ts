import {isEqual} from 'lodash-es';
import {DEFAULT_THEME} from '../constants.js';
import type {AnyPrivateColorToken} from '../private-colors/types.js';
import type {GenerateOptions, GravityTheme, Theme, UtilityColor} from '../types.js';
import {
    TEXT_GROUP_PROPERTIES,
    TEXT_VARIANT_PROPERTIES,
    type TextGroup,
    type TextVariant,
} from '../typography/types.js';
import {
    createFontCssVariable,
    createTextCssVariable,
    generateCssFontFamily,
} from '../typography/utils.js';
import {
    createPrivateColorCssVariable,
    createUtilityColorCssVariable,
    isInternalPrivateColorReference,
    isInternalUtilityColorReference,
    parseInternalPrivateColorReference,
    parseInternalUtilityColorReference,
} from '../utils.js';
import type {BorderSize} from '../borders/types.js';
import {createBorderRadiusCssVariable} from '../borders/utils.js';

/**
 * Adds indentation to each line of a string
 * @param text - Text to indent
 * @param spaces - Number of spaces for indentation
 * @returns Indented text
 */
const addIndentation = (text: string, spaces: number = 4): string => {
    const indent = ' '.repeat(spaces);
    return text
        .split('\n')
        .map((line) => (line.trim() ? indent + line : line))
        .join('\n');
};

const COMMON_VARIABLES_TEMPLATE_NAME = '%COMMON_VARIABLES%';
const LIGHT_THEME_VARIABLES_TEMPLATE_NAME = '%LIGHT_THEME_VARIABLES%';
const DARK_THEME_VARIABLES_TEMPLATE_NAME = '%DARK_THEME_VARIABLES%';

const CSS_TEMPLATE = `
.g-root {
${COMMON_VARIABLES_TEMPLATE_NAME}
}

.g-root_theme_light {
${LIGHT_THEME_VARIABLES_TEMPLATE_NAME}
}

.g-root_theme_dark {
${DARK_THEME_VARIABLES_TEMPLATE_NAME}
}
`.trim();

const isBackgroundColorChanged = (theme: GravityTheme) => {
    return (
        !isEqual(
            DEFAULT_THEME.utilityColors['base-background'].dark,
            theme.utilityColors['base-background'].dark,
        ) ||
        !isEqual(
            DEFAULT_THEME.utilityColors['base-background'].light,
            theme.utilityColors['base-background'].light,
        )
    );
};

const createUtilityColorExport = (
    theme: GravityTheme,
    themeVariant: Theme,
    utilityColor: UtilityColor,
    forPreview?: boolean,
) => {
    const {ref, value} = theme.utilityColors[utilityColor][themeVariant];
    let resultValue = ref ? ref : value;

    if (isInternalPrivateColorReference(resultValue)) {
        const parseResult = parseInternalPrivateColorReference(resultValue);
        if (parseResult) {
            const {mainColorToken, privateColorCode} = parseResult;
            resultValue = `var(${createPrivateColorCssVariable(mainColorToken, privateColorCode)})`;
        }
    } else if (isInternalUtilityColorReference(resultValue)) {
        const utilityColor = parseInternalUtilityColorReference(resultValue);
        if (utilityColor) {
            resultValue = `var(${createUtilityColorCssVariable(utilityColor)})`;
        }
    }

    return [
        createUtilityColorCssVariable(utilityColor),
        `${resultValue}${forPreview ? ' !important' : ''};`,
    ].join(': ');
};

/*
Output example:

.g-root_theme_light {
    --g-color-private-brand-50: rgba(203,255,92,0.1);
    --g-color-base-brand-hover: var(--g-color-private-brand-600-solid);
}

.g-root_theme_dark {
    --g-color-private-brand-50: rgba(203,255,92,0.1);
    --g-color-base-brand-hover: var(--g-color-private-brand-650-solid);
}
*/
/**
 * Generates CSS string from a GravityTheme object.
 * @param options - The options for generating CSS
 * @returns CSS string representation of the theme
 */
export function generateCSS({theme, ignoreDefaultValues, forPreview}: GenerateOptions): string {
    const backgroundColorChanged = isBackgroundColorChanged(theme);

    const prepareThemeVariables = (themeVariant: Theme) => {
        let cssVariables = '';

        Object.entries(theme.baseColors).forEach(([token, colorOptions]) => {
            // Dont export colors that are equals to default
            // Private colors recalculate when background color changes
            const valueEqualsToDefault =
                DEFAULT_THEME.baseColors[token]?.[themeVariant].value ===
                    colorOptions[themeVariant].value && !backgroundColorChanged;

            if (valueEqualsToDefault && ignoreDefaultValues) {
                return;
            }

            if (theme.privateColors[token]?.[themeVariant]) {
                Object.entries(theme.privateColors[token][themeVariant]).forEach(
                    ([privateColorToken, color]) => {
                        const valueInDefaultTheme =
                            DEFAULT_THEME.privateColors?.[token]?.[themeVariant]?.[
                                privateColorToken as AnyPrivateColorToken
                            ]?.value;

                        const colorEqualsToDefault =
                            valueInDefaultTheme === color.value &&
                            valueInDefaultTheme !== undefined;

                        if (ignoreDefaultValues && colorEqualsToDefault) {
                            return;
                        }

                        cssVariables += `${createPrivateColorCssVariable(
                            token,
                            privateColorToken as AnyPrivateColorToken,
                        )}: ${color.value}${forPreview ? ' !important' : ''};\n`;
                    },
                );
                cssVariables += '\n';
            }
        });

        cssVariables += '\n';

        Object.keys(theme.utilityColors).forEach((_utilityColor) => {
            const utilityColor = _utilityColor as UtilityColor;

            // Dont export colors that are equals to default
            const valueEqualsToDefault =
                DEFAULT_THEME.utilityColors[utilityColor][themeVariant].value ===
                    theme.utilityColors[utilityColor][themeVariant].value ||
                (DEFAULT_THEME.utilityColors[utilityColor][themeVariant].ref &&
                    DEFAULT_THEME.utilityColors[utilityColor][themeVariant].ref ===
                        theme.utilityColors[utilityColor][themeVariant].ref);

            if (valueEqualsToDefault && ignoreDefaultValues) {
                return;
            }

            cssVariables += `${createUtilityColorExport(theme, themeVariant, utilityColor, forPreview)}\n`;
        });

        return cssVariables.trim();
    };

    const prepateBordersVariables = () => {
        let cssVariables = '';

        for (const [size, value] of Object.entries(theme.borders)) {
            if (ignoreDefaultValues && isEqual(DEFAULT_THEME.borders[size as BorderSize], value)) {
                continue;
            }

            const entry = [
                createBorderRadiusCssVariable(size as BorderSize),
                `${value}${forPreview ? ' !important' : ''};`,
            ].join(': ');
            cssVariables += entry + '\n';
        }

        return cssVariables;
    };

    const prepareTypographyVariables = () => {
        let cssVariables = '';

        for (const [key, value] of Object.entries(theme.typography.fontFamilies)) {
            if (ignoreDefaultValues && isEqual(DEFAULT_THEME.typography.fontFamilies[key], value)) {
                continue;
            }

            if (!value) {
                continue;
            }

            const entry = [
                createFontCssVariable(key),
                `${generateCssFontFamily(value)}${forPreview ? ' !important' : ''};`,
            ].join(': ');
            cssVariables += entry + '\n';
        }

        for (const [group, value] of Object.entries(theme.typography.groups)) {
            for (const property of TEXT_GROUP_PROPERTIES) {
                const cssVariable = createTextCssVariable({
                    group: group as TextGroup,
                    property,
                });

                if (
                    ignoreDefaultValues &&
                    isEqual(
                        DEFAULT_THEME.typography.groups[group as TextGroup][property],
                        value[property],
                    )
                ) {
                    continue;
                }

                if (property === 'font-family') {
                    const val = value[property];
                    const refCssVariable = createFontCssVariable(val);
                    const entry = [
                        cssVariable,
                        `var(${refCssVariable})${forPreview ? ' !important' : ''};`,
                    ].join(': ');
                    cssVariables += entry + '\n';
                } else {
                    const entry = [
                        cssVariable,
                        `${value[property]}${forPreview ? ' !important' : ''};`,
                    ].join(': ');
                    cssVariables += entry + '\n';
                }
            }
        }

        for (const [variant, value] of Object.entries(theme.typography.variants)) {
            for (const property of TEXT_VARIANT_PROPERTIES) {
                if (
                    ignoreDefaultValues &&
                    isEqual(
                        DEFAULT_THEME.typography.variants[variant as TextVariant][property],
                        value[property],
                    )
                ) {
                    continue;
                }

                const cssVariable = createTextCssVariable({
                    property,
                    variant: variant as TextVariant,
                });
                const entry = [
                    cssVariable,
                    `${value[property]}${forPreview ? ' !important' : ''};`,
                ].join(': ');
                cssVariables += entry + '\n';
            }
        }

        return cssVariables.trim();
    };

    const prepareCommonVariables = () => {
        return [prepateBordersVariables(), prepareTypographyVariables()].join('\n').trim();
    };

    const themeParams = {
        common: prepareCommonVariables(),
        light: prepareThemeVariables('light'),
        dark: prepareThemeVariables('dark'),
    };

    return CSS_TEMPLATE.replace(COMMON_VARIABLES_TEMPLATE_NAME, addIndentation(themeParams.common))
        .replace(LIGHT_THEME_VARIABLES_TEMPLATE_NAME, addIndentation(themeParams.light))
        .replace(DARK_THEME_VARIABLES_TEMPLATE_NAME, addIndentation(themeParams.dark))
        .replace(/\n{3,}/g, '\n\n');
}
