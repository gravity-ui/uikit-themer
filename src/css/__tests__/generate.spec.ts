import {cloneDeep} from 'lodash-es';
import {DEFAULT_THEME} from '../../constants.js';
import {generateCSS} from '../generate.js';
import {
    updateUtilityColor,
    updateBaseColor,
    createInternalPrivateColorReference,
    createInternalUtilityColorReference,
} from '../../utils.js';

describe('generateCSS', () => {
    it('generate full theme', async () => {
        const result = generateCSS({theme: DEFAULT_THEME});
        expect(result).toMatchSnapshot();
    });

    it('generate only changed variables', async () => {
        const newTheme = cloneDeep(DEFAULT_THEME);
        newTheme.typography.variants['body-1']['font-size'] = '200px';
        newTheme.utilityColors['text-link-hover'] = {
            light: {
                value: 'red',
            },
            dark: {
                value: 'blue',
            },
        };
        const result = generateCSS({theme: newTheme, ignoreDefaultValues: true});
        expect(result).toMatchSnapshot();
    });

    it('generate when add new base color', async () => {
        let newTheme = cloneDeep(DEFAULT_THEME);
        newTheme = updateBaseColor({
            theme: newTheme,
            colorToken: 'brand',
            themeVariant: 'light',
            value: 'rgb(233, 233, 233)',
        });

        newTheme = updateBaseColor({
            theme: newTheme,
            colorToken: 'custom',
            value: {
                light: 'rgb(255,255,255)',
                dark: 'rgb(253,255,23)',
            },
        });
        const result = generateCSS({theme: newTheme, ignoreDefaultValues: true});
        expect(result).toMatchSnapshot();
    });

    it('generate when set reference in utility colors', async () => {
        let newTheme = cloneDeep(DEFAULT_THEME);
        newTheme = updateUtilityColor({
            theme: newTheme,
            colorToken: 'text-link-hover',
            themeVariant: 'light',
            value: createInternalPrivateColorReference('red', '500-solid'),
        });

        newTheme = updateUtilityColor({
            theme: newTheme,
            colorToken: 'base-float-medium',
            themeVariant: 'dark',
            value: createInternalUtilityColorReference('base-background'),
        });

        const result = generateCSS({theme: newTheme, ignoreDefaultValues: true});
        expect(result).toMatchSnapshot();
    });
});
