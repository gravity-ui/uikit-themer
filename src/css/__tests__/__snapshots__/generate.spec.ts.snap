// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateCSS generate full theme 1`] = `
".g-root {
--g-border-radius-xs: 3px;
--g-border-radius-s: 5px;
--g-border-radius-m: 6px;
--g-border-radius-l: 8px;
--g-border-radius-xl: 10px;

--g-font-family-sans: 'Inter', 'Helvetica Neue', 'Helvetica', 'Arial';
--g-font-family-monospace: 'Menlo', 'Monaco', 'Consolas', 'Liberation Mono', 'Ubuntu Mono', 'DejaVu Sans Mono', 'Courier New', 'Courier';
--g-text-body-font-weight: 400;
--g-text-body-font-family: var(--g-font-family-sans);
--g-text-caption-font-weight: 400;
--g-text-caption-font-family: var(--g-font-family-sans);
--g-text-code-font-weight: 400;
--g-text-code-font-family: var(--g-font-family-monospace);
--g-text-display-font-weight: 600;
--g-text-display-font-family: var(--g-font-family-sans);
--g-text-header-font-weight: 600;
--g-text-header-font-family: var(--g-font-family-sans);
--g-text-subheader-font-weight: 600;
--g-text-subheader-font-family: var(--g-font-family-sans);
--g-text-body-short-font-size: 13px;
--g-text-body-short-line-height: 16px;
--g-text-body-1-font-size: 13px;
--g-text-body-1-line-height: 18px;
--g-text-body-2-font-size: 15px;
--g-text-body-2-line-height: 20px;
--g-text-body-3-font-size: 17px;
--g-text-body-3-line-height: 24px;
--g-text-caption-1-font-size: 9px;
--g-text-caption-1-line-height: 12px;
--g-text-caption-2-font-size: 11px;
--g-text-caption-2-line-height: 16px;
--g-text-code-1-font-size: 12px;
--g-text-code-1-line-height: 18px;
--g-text-code-inline-1-font-size: 12px;
--g-text-code-inline-1-line-height: 14px;
--g-text-code-2-font-size: 14px;
--g-text-code-2-line-height: 20px;
--g-text-code-inline-2-font-size: 14px;
--g-text-code-inline-2-line-height: 16px;
--g-text-code-3-font-size: 16px;
--g-text-code-3-line-height: 24px;
--g-text-code-inline-3-font-size: 16px;
--g-text-code-inline-3-line-height: 20px;
--g-text-display-1-font-size: 28px;
--g-text-display-1-line-height: 36px;
--g-text-display-2-font-size: 32px;
--g-text-display-2-line-height: 40px;
--g-text-display-3-font-size: 40px;
--g-text-display-3-line-height: 48px;
--g-text-display-4-font-size: 48px;
--g-text-display-4-line-height: 52px;
--g-text-header-1-font-size: 20px;
--g-text-header-1-line-height: 24px;
--g-text-header-2-font-size: 24px;
--g-text-header-2-line-height: 28px;
--g-text-subheader-1-font-size: 13px;
--g-text-subheader-1-line-height: 18px;
--g-text-subheader-2-font-size: 15px;
--g-text-subheader-2-line-height: 20px;
--g-text-subheader-3-font-size: 17px;
--g-text-subheader-3-line-height: 24px;
}

.g-root_theme_light {
--g-color-private-brand-50: rgb(203 255 92 / 0.1);
--g-color-private-brand-100: rgb(203 255 92 / 0.15);
--g-color-private-brand-150: rgb(203 255 92 / 0.2);
--g-color-private-brand-200: rgb(203 255 92 / 0.3);
--g-color-private-brand-250: rgb(203 255 92 / 0.4);
--g-color-private-brand-300: rgb(203 255 92 / 0.5);
--g-color-private-brand-350: rgb(203 255 92 / 0.6);
--g-color-private-brand-400: rgb(203 255 92 / 0.7);
--g-color-private-brand-450: rgb(203 255 92 / 0.8);
--g-color-private-brand-500: rgb(203 255 92 / 0.9);
--g-color-private-brand-550-solid: rgb(203 255 92);
--g-color-private-brand-1000-solid: rgb(59 63 43);
--g-color-private-brand-950-solid: rgb(68 74 46);
--g-color-private-brand-900-solid: rgb(85 97 51);
--g-color-private-brand-850-solid: rgb(102 119 57);
--g-color-private-brand-800-solid: rgb(119 142 63);
--g-color-private-brand-750-solid: rgb(135 165 69);
--g-color-private-brand-700-solid: rgb(152 187 75);
--g-color-private-brand-650-solid: rgb(169 210 80);
--g-color-private-brand-600-solid: rgb(186 232 86);
--g-color-private-brand-500-solid: rgb(208 255 108);
--g-color-private-brand-450-solid: rgb(213 255 125);
--g-color-private-brand-400-solid: rgb(219 255 141);
--g-color-private-brand-350-solid: rgb(224 255 157);
--g-color-private-brand-300-solid: rgb(229 255 174);
--g-color-private-brand-250-solid: rgb(234 255 190);
--g-color-private-brand-200-solid: rgb(239 255 206);
--g-color-private-brand-150-solid: rgb(245 255 222);
--g-color-private-brand-100-solid: rgb(247 255 231);
--g-color-private-brand-50-solid: rgb(250 255 239);


--g-color-base-brand: rgb(203,255,92);
--g-color-base-background: rgb(255,255,255);
--g-color-base-brand-hover: var(--g-color-private-brand-600-solid);
--g-color-base-selection: var(--g-color-private-brand-200);
--g-color-base-selection-hover: var(--g-color-private-brand-300);
--g-color-line-brand: var(--g-color-private-brand-600-solid);
--g-color-text-brand: var(--g-color-private-brand-700-solid);
--g-color-text-brand-heavy: var(--g-color-private-brand-700-solid);
--g-color-text-brand-contrast: rgba(0, 0, 0, 0.85);
--g-color-text-link: var(--g-color-private-brand-600-solid);
--g-color-text-link-hover: var(--g-color-private-brand-800-solid);
--g-color-text-link-visited: var(--g-color-private-purple-550-solid);
--g-color-text-link-visited-hover: var(--g-color-private-purple-800-solid);
}

.g-root_theme_dark {
--g-color-private-brand-50: rgb(203 255 92 / 0.1);
--g-color-private-brand-100: rgb(203 255 92 / 0.15);
--g-color-private-brand-150: rgb(203 255 92 / 0.2);
--g-color-private-brand-200: rgb(203 255 92 / 0.3);
--g-color-private-brand-250: rgb(203 255 92 / 0.4);
--g-color-private-brand-300: rgb(203 255 92 / 0.5);
--g-color-private-brand-350: rgb(203 255 92 / 0.6);
--g-color-private-brand-400: rgb(203 255 92 / 0.7);
--g-color-private-brand-450: rgb(203 255 92 / 0.8);
--g-color-private-brand-500: rgb(203 255 92 / 0.9);
--g-color-private-brand-550-solid: rgb(203 255 92);
--g-color-private-brand-1000-solid: rgb(247 255 231);
--g-color-private-brand-950-solid: rgb(245 255 222);
--g-color-private-brand-900-solid: rgb(239 255 206);
--g-color-private-brand-850-solid: rgb(234 255 190);
--g-color-private-brand-800-solid: rgb(229 255 174);
--g-color-private-brand-750-solid: rgb(224 255 157);
--g-color-private-brand-700-solid: rgb(219 255 141);
--g-color-private-brand-650-solid: rgb(213 255 125);
--g-color-private-brand-600-solid: rgb(208 255 108);
--g-color-private-brand-500-solid: rgb(186 232 86);
--g-color-private-brand-450-solid: rgb(169 210 80);
--g-color-private-brand-400-solid: rgb(152 187 75);
--g-color-private-brand-350-solid: rgb(135 165 69);
--g-color-private-brand-300-solid: rgb(119 142 63);
--g-color-private-brand-250-solid: rgb(102 119 57);
--g-color-private-brand-200-solid: rgb(85 97 51);
--g-color-private-brand-150-solid: rgb(68 74 46);
--g-color-private-brand-100-solid: rgb(59 63 43);
--g-color-private-brand-50-solid: rgb(51 52 40);


--g-color-base-brand: rgb(203,255,92);
--g-color-base-background: rgb(34,29,34);
--g-color-base-brand-hover: var(--g-color-private-brand-650-solid);
--g-color-base-selection: var(--g-color-private-brand-150);
--g-color-base-selection-hover: var(--g-color-private-brand-200);
--g-color-line-brand: var(--g-color-private-brand-600-solid);
--g-color-text-brand: var(--g-color-private-brand-600-solid);
--g-color-text-brand-heavy: var(--g-color-private-brand-700-solid);
--g-color-text-brand-contrast: rgba(0, 0, 0, 0.9);
--g-color-text-link: var(--g-color-private-brand-550-solid);
--g-color-text-link-hover: var(--g-color-private-brand-700-solid);
--g-color-text-link-visited: var(--g-color-private-purple-700-solid);
--g-color-text-link-visited-hover: var(--g-color-private-purple-850-solid);
}"
`;

exports[`generateCSS generate only changed variables 1`] = `
".g-root {
--g-text-body-1-font-size: 200px;
}

.g-root_theme_light {
--g-color-private-brand-50: rgb(203 255 92 / 0.1);
--g-color-private-brand-100: rgb(203 255 92 / 0.15);
--g-color-private-brand-150: rgb(203 255 92 / 0.2);
--g-color-private-brand-200: rgb(203 255 92 / 0.3);
--g-color-private-brand-250: rgb(203 255 92 / 0.4);
--g-color-private-brand-300: rgb(203 255 92 / 0.5);
--g-color-private-brand-350: rgb(203 255 92 / 0.6);
--g-color-private-brand-400: rgb(203 255 92 / 0.7);
--g-color-private-brand-450: rgb(203 255 92 / 0.8);
--g-color-private-brand-500: rgb(203 255 92 / 0.9);
--g-color-private-brand-550-solid: rgb(203 255 92);
--g-color-private-brand-1000-solid: rgb(59 63 43);
--g-color-private-brand-950-solid: rgb(68 74 46);
--g-color-private-brand-900-solid: rgb(85 97 51);
--g-color-private-brand-850-solid: rgb(102 119 57);
--g-color-private-brand-800-solid: rgb(119 142 63);
--g-color-private-brand-750-solid: rgb(135 165 69);
--g-color-private-brand-700-solid: rgb(152 187 75);
--g-color-private-brand-650-solid: rgb(169 210 80);
--g-color-private-brand-600-solid: rgb(186 232 86);
--g-color-private-brand-500-solid: rgb(208 255 108);
--g-color-private-brand-450-solid: rgb(213 255 125);
--g-color-private-brand-400-solid: rgb(219 255 141);
--g-color-private-brand-350-solid: rgb(224 255 157);
--g-color-private-brand-300-solid: rgb(229 255 174);
--g-color-private-brand-250-solid: rgb(234 255 190);
--g-color-private-brand-200-solid: rgb(239 255 206);
--g-color-private-brand-150-solid: rgb(245 255 222);
--g-color-private-brand-100-solid: rgb(247 255 231);
--g-color-private-brand-50-solid: rgb(250 255 239);


--g-color-base-brand: rgb(203,255,92);
--g-color-base-background: rgb(255,255,255);
--g-color-base-brand-hover: var(--g-color-private-brand-600-solid);
--g-color-base-selection: var(--g-color-private-brand-200);
--g-color-base-selection-hover: var(--g-color-private-brand-300);
--g-color-line-brand: var(--g-color-private-brand-600-solid);
--g-color-text-brand: var(--g-color-private-brand-700-solid);
--g-color-text-brand-heavy: var(--g-color-private-brand-700-solid);
--g-color-text-brand-contrast: rgba(0, 0, 0, 0.85);
--g-color-text-link: var(--g-color-private-brand-600-solid);
--g-color-text-link-hover: var(--g-color-private-brand-800-solid);
--g-color-text-link-visited: var(--g-color-private-purple-550-solid);
--g-color-text-link-visited-hover: var(--g-color-private-purple-800-solid);
}

.g-root_theme_dark {
--g-color-private-brand-50: rgb(203 255 92 / 0.1);
--g-color-private-brand-100: rgb(203 255 92 / 0.15);
--g-color-private-brand-150: rgb(203 255 92 / 0.2);
--g-color-private-brand-200: rgb(203 255 92 / 0.3);
--g-color-private-brand-250: rgb(203 255 92 / 0.4);
--g-color-private-brand-300: rgb(203 255 92 / 0.5);
--g-color-private-brand-350: rgb(203 255 92 / 0.6);
--g-color-private-brand-400: rgb(203 255 92 / 0.7);
--g-color-private-brand-450: rgb(203 255 92 / 0.8);
--g-color-private-brand-500: rgb(203 255 92 / 0.9);
--g-color-private-brand-550-solid: rgb(203 255 92);
--g-color-private-brand-1000-solid: rgb(247 255 231);
--g-color-private-brand-950-solid: rgb(245 255 222);
--g-color-private-brand-900-solid: rgb(239 255 206);
--g-color-private-brand-850-solid: rgb(234 255 190);
--g-color-private-brand-800-solid: rgb(229 255 174);
--g-color-private-brand-750-solid: rgb(224 255 157);
--g-color-private-brand-700-solid: rgb(219 255 141);
--g-color-private-brand-650-solid: rgb(213 255 125);
--g-color-private-brand-600-solid: rgb(208 255 108);
--g-color-private-brand-500-solid: rgb(186 232 86);
--g-color-private-brand-450-solid: rgb(169 210 80);
--g-color-private-brand-400-solid: rgb(152 187 75);
--g-color-private-brand-350-solid: rgb(135 165 69);
--g-color-private-brand-300-solid: rgb(119 142 63);
--g-color-private-brand-250-solid: rgb(102 119 57);
--g-color-private-brand-200-solid: rgb(85 97 51);
--g-color-private-brand-150-solid: rgb(68 74 46);
--g-color-private-brand-100-solid: rgb(59 63 43);
--g-color-private-brand-50-solid: rgb(51 52 40);


--g-color-base-brand: rgb(203,255,92);
--g-color-base-background: rgb(34,29,34);
--g-color-base-brand-hover: var(--g-color-private-brand-650-solid);
--g-color-base-selection: var(--g-color-private-brand-150);
--g-color-base-selection-hover: var(--g-color-private-brand-200);
--g-color-line-brand: var(--g-color-private-brand-600-solid);
--g-color-text-brand: var(--g-color-private-brand-600-solid);
--g-color-text-brand-heavy: var(--g-color-private-brand-700-solid);
--g-color-text-brand-contrast: rgba(0, 0, 0, 0.9);
--g-color-text-link: var(--g-color-private-brand-550-solid);
--g-color-text-link-hover: var(--g-color-private-brand-700-solid);
--g-color-text-link-visited: var(--g-color-private-purple-700-solid);
--g-color-text-link-visited-hover: var(--g-color-private-purple-850-solid);
}"
`;
